src/compiler/checker.ts

```diff
@@ -6691,19 +6746,35 @@ namespace ts {
                 }
                 const signature = getResolvedSignature(expr);
 
-                if (signature.typePredicate &&
-                    expr.arguments[signature.typePredicate.parameterIndex] &&
-                    getSymbolAtLocation(expr.arguments[signature.typePredicate.parameterIndex]) === symbol) {
+                if (!signature.typePredicate) {
+                    return type;
+                }
+                const predicate = signature.typePredicate;
+                if (isIdentifierTypePredicate(predicate)) {
+                    if (expr.arguments[predicate.parameterIndex] &&
+                        getSymbolAtLocation(expr.arguments[predicate.parameterIndex]) === symbol) {

```

Comment:
But the `getSymbolAtLocation` function is intended only for the language service and has fuzzy semantics that aren't part of the language. It's definitely not the right function to use here. I know this code is not part of the PR, but we should fix it.